File: cpp_sdk/cpp_dist_bundle/benchmarks/benchmark.cpp
// The following code runs speed benchmarks for the different modules
// The first few inferences are discarded to ensure caching is hot

#include "benchmark.h"
#include "observation.h"
#include "sdkfactory.h"
#include "stopwatch.h"

#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "tf_sdk.h"

using namespace Trueface;

int main() {
    uint32_t batchSize = 16;
    GPUOptions gpuOptions = Benchmarks::SDKFactory::createGPUOptions(
        false,     // enableGPU,  NOTE: set this to true to benchmark on GPU
        0,         // deviceIndex
        batchSize, // maxBatchSize
        1          // optBatchSize
    );

    std::cout << "==========================" << std::endl;
    std::cout << "==========================" << std::endl;
    if (gpuOptions.enableGPU) {
        std::cout << "Using GPU for inference" << std::endl;
    } else {
        std::cout << "Using CPU for inference" << std::endl;
    }
    std::cout << "==========================" << std::endl;
    std::cout << "==========================" << std::endl;

    unsigned int multFactor = 1;
    if (gpuOptions.enableGPU) {
        multFactor = 10;
    }

    bool warmup = true; // Warmup inference to ensure caching is hot
    int numWarmup = 10;
    Benchmarks::SDKFactory sdkFactory(gpuOptions);
    Benchmarks::ObservationList observations;

    benchmarkPreprocessImage(sdkFactory, {warmup, numWarmup, 1, 200}, observations);

    benchmarkFaceLandmarkDetection(sdkFactory, {warmup, numWarmup, 1, 100 * multFactor}, observations);

    benchmarkHeadOrientation(sdkFactory, {warmup, numWarmup, 1, 500 * multFactor}, observations);

    benchmarkGlassesDetection(sdkFactory, {warmup, numWarmup, 1, 200 * multFactor}, observations);

    benchmarkSpoofDetection(sdkFactory, {warmup, numWarmup, 1, 100 * multFactor}, observations);

    benchmarkObjectDetection(sdkFactory, ObjectDetectionModel::FAST, {warmup, numWarmup, 1, 100 * multFactor}, observations);

    benchmarkObjectDetection(sdkFactory, ObjectDetectionModel::ACCURATE, {warmup, numWarmup, 1, 40 * multFactor}, observations);

    Benchmarks::Parameters frBenchmarkParams{warmup, numWarmup, 1, 40 * multFactor};
    std::vector<unsigned int> batchSizes;
    if (gpuOptions.enableGPU) {
        // Only test with batching when GPU is enabled.
        // Batching is not supported by CPU and will not cause a speedup.
        batchSizes = {1, batchSize};
    } else {
        batchSizes = {1};
    }
    // Methods which support batch inference
    for (auto currentBatchSize : batchSizes) {
        frBenchmarkParams.batchSize = currentBatchSize;
        benchmarkFaceImageOrientationDetection(sdkFactory, {warmup, numWarmup, currentBatchSize, 50 * multFactor}, observations);

        benchmarkDetailedLandmarkDetection(sdkFactory, {warmup, numWarmup, currentBatchSize, 100 * multFactor}, observations);

        benchmarkFaceImageBlurDetection(sdkFactory, {warmup, numWarmup, currentBatchSize, 200 * multFactor}, observations);

        benchmarkBlinkDetection(sdkFactory, {warmup, numWarmup, currentBatchSize, 100 * multFactor}, observations);

        benchmarkMaskDetection(sdkFactory, {warmup, numWarmup, currentBatchSize, 200 * multFactor}, observations);

        benchmarkFaceTemplateQualityEstimation(sdkFactory,  {warmup, numWarmup, currentBatchSize, 200 * multFactor}, observations);

        benchmarkFaceRecognition(sdkFactory, FacialRecognitionModel::LITE_V2, frBenchmarkParams, observations);

        benchmarkFaceRecognition(sdkFactory, FacialRecognitionModel::LITE_V3, frBenchmarkParams, observations);

        benchmarkFaceRecognition(sdkFactory, FacialRecognitionModel::TFV5_2, frBenchmarkParams, observations);

        benchmarkFaceRecognition(sdkFactory, FacialRecognitionModel::TFV6, frBenchmarkParams, observations);

        benchmarkFaceRecognition(sdkFactory, FacialRecognitionModel::TFV7, frBenchmarkParams, observations);
    }

    Benchmarks::ObservationCSVWriter csv{"benchmarks.csv"};
    csv.write(observations);

    return EXIT_SUCCESS;
}

File: python/python_dist_bundle/benchmarks/benchmark.py
# Run benchmarks

import preprocess_image
import face_image_orientation_detection
import face_landmark_detection
import detailed_landmark_detection
import head_orientation
import face_image_blur_detection
import blink_detection
import mask_detection
import glasses_detection
import spoof_detection
import object_detection
import face_recognition
import face_template_quality_estimation
from observation import (Observation, ObservationCSVWriter)
from typing import List
from utils import Parameters


import tfsdk


def main() -> None:
    NUM_WARMUP = 10
    DO_WARMUP = True

    gpu_options = tfsdk.GPUOptions()
    gpu_options.enable_GPU = False  # TODO: Set this to true to benchmark on GPU.
    gpu_options.device_index = 0

    gpu_module_options = tfsdk.GPUModuleOptions()
    gpu_module_options.precision = tfsdk.PRECISION.FP16

    batch_size = 16
    gpu_module_options.max_batch_size = batch_size
    gpu_module_options.opt_batch_size = 1

    gpu_options.face_detector_GPU_options = gpu_module_options
    gpu_options.face_recognizer_GPU_options = gpu_module_options
    gpu_options.mask_detector_GPU_options = gpu_module_options
    gpu_options.object_detector_GPU_options = gpu_module_options
    gpu_options.face_landmark_detector_GPU_options = gpu_module_options
    gpu_options.face_orientation_detector_GPU_options = gpu_module_options
    gpu_options.face_blur_detector_GPU_options = gpu_module_options
    gpu_options.spoof_detector_GPU_options = gpu_module_options
    gpu_options.blink_detector_GPU_options = gpu_module_options
    gpu_options.face_template_quality_estimator_GPU_options = gpu_module_options

    print("==========================")
    print("==========================")
    if gpu_options.enable_GPU is True:
        print("Using GPU for inference")
    else:
        print("Using CPU for inference")
    print("==========================")
    print("==========================")

    mult_factor = 1
    if gpu_options.enable_GPU is True:
        mult_factor = 10

    observations: List[Observation] = []
    parameters = Parameters(DO_WARMUP, NUM_WARMUP, 1, 200)
    preprocess_image.benchmark(gpu_options, parameters, observations)
    glasses_detection.benchmark(gpu_options, parameters, observations)

    parameters.num_iterations = 50 * mult_factor
    face_image_orientation_detection.benchmark(gpu_options, parameters, observations)

    parameters.num_iterations = 100 * mult_factor
    face_landmark_detection.benchmark(gpu_options, parameters, observations)
    detailed_landmark_detection.benchmark(gpu_options, parameters, observations)
    blink_detection.benchmark(gpu_options, parameters, observations)
    mask_detection.benchmark(gpu_options, parameters, observations)
    spoof_detection.benchmark(gpu_options, parameters, observations)
    face_template_quality_estimation.benchmark(gpu_options, parameters, observations)

    parameters.num_iterations = 200 * mult_factor
    face_image_blur_detection.benchmark(gpu_options, parameters, observations)

    parameters.num_iterations = 500 * mult_factor
    head_orientation.benchmark(gpu_options, parameters, observations)

    parameters.num_iterations = 50 * mult_factor
    for object_detection_model in (tfsdk.OBJECTDETECTIONMODEL.FAST,
                                   tfsdk.OBJECTDETECTIONMODEL.ACCURATE):
        object_detection.benchmark(gpu_options, object_detection_model, parameters, observations)

    parameters.num_iterations = 200
    face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.LITE_V2, gpu_options, parameters, observations)
    face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.LITE_V3, gpu_options, parameters, observations)

    parameters.num_iterations = 40 * mult_factor
    face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.TFV5_2, gpu_options, parameters, observations)
    face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.TFV6, gpu_options, parameters, observations)
    face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.TFV7, gpu_options, parameters, observations)
    # Benchmarks with batching.
    # Only run for GPU inference
    # CPU inference does not support batching, so will provide no speedup
    if gpu_options.enable_GPU:
        parameters.batch_size = batch_size
        face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.TFV5_2, gpu_options, parameters, observations)
        face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.TFV6, gpu_options, parameters, observations)
        face_recognition.benchmark(tfsdk.FACIALRECOGNITIONMODEL.TFV7, gpu_options, parameters, observations)

        parameters.num_iterations = 100 * mult_factor
        mask_detection.benchmark(gpu_options, parameters, observations)
        face_template_quality_estimation.benchmark(gpu_options, parameters, observations)
        detailed_landmark_detection.benchmark(gpu_options, parameters, observations)
        blink_detection.benchmark(gpu_options, parameters, observations)
        face_image_orientation_detection.benchmark(gpu_options, parameters, observations)
        face_image_blur_detection.benchmark(gpu_options, parameters, observations)


    ObservationCSVWriter('benchmarks.csv').write(observations)


if __name__ == '__main__':
    main()
